pinot-core/src/main/java/org/apache/pinot/core/transport/ImplicitHybridTableRouteInfo.java

```diff
@@ -0,0 +1,135 @@
+      Map<ServerInstance, ServerRouteInfo> offlineRoutingTable,
+      Map<ServerInstance, ServerRouteInfo> realtimeRoutingTable) {
+    _offlineBrokerRequest = offlineBrokerRequest;
+    _realtimeBrokerRequest = realtimeBrokerRequest;
+    _offlineRoutingTable = offlineRoutingTable;
+    _realtimeRoutingTable = realtimeRoutingTable;
+  }
+
+  @Nullable
+  @Override
+  public BrokerRequest getOfflineBrokerRequest() {
+    return _offlineBrokerRequest;
+  }
+
+  @Nullable
+  @Override
+  public BrokerRequest getRealtimeBrokerRequest() {
+    return _realtimeBrokerRequest;
+  }
+
+  @Nullable
+  @Override
+  public Map<ServerInstance, ServerRouteInfo> getOfflineRoutingTable() {
+    return _offlineRoutingTable;
+  }
+
+  @Nullable
+  @Override
+  public Map<ServerInstance, ServerRouteInfo> getRealtimeRoutingTable() {
+    return _realtimeRoutingTable;
+  }
+
+  protected static Map<ServerRoutingInstance, InstanceRequest> getRequestMapFromRoutingTable(TableType tableType,
+      Map<ServerInstance, ServerRouteInfo> routingTable, BrokerRequest brokerRequest, long requestId, String brokerId,
+      boolean preferTls) {
+    Map<ServerRoutingInstance, InstanceRequest> requestMap = new HashMap<>();
+    for (Map.Entry<ServerInstance, ServerRouteInfo> entry : routingTable.entrySet()) {
+      ServerRoutingInstance serverRoutingInstance = entry.getKey().toServerRoutingInstance(tableType, preferTls);
+      InstanceRequest instanceRequest = getInstanceRequest(requestId, brokerId, brokerRequest, entry.getValue());
+      requestMap.put(serverRoutingInstance, instanceRequest);
+    }
+    return requestMap;
+  }
+
+  protected static InstanceRequest getInstanceRequest(long requestId, String brokerId, BrokerRequest brokerRequest,
+      ServerRouteInfo segments) {
+    InstanceRequest instanceRequest = new InstanceRequest();
+    instanceRequest.setRequestId(requestId);
+    instanceRequest.setCid(QueryThreadContext.getCid());
+    instanceRequest.setQuery(brokerRequest);
+    Map<String, String> queryOptions = brokerRequest.getPinotQuery().getQueryOptions();
+    if (queryOptions != null) {
+      instanceRequest.setEnableTrace(Boolean.parseBoolean(queryOptions.get(CommonConstants.Broker.Request.TRACE)));
+    }
+    instanceRequest.setSearchSegments(segments.getSegments());
+    instanceRequest.setBrokerId(brokerId);
+    if (CollectionUtils.isNotEmpty(segments.getOptionalSegments())) {
+      // Don't set this field, i.e. leave it as null, if there is no optional segment at all, to be more backward
+      // compatible, as there are places like in multi-stage query engine where this field is not set today when
+      // creating the InstanceRequest.
+      instanceRequest.setOptionalSegments(segments.getOptionalSegments());
+    }
+    return instanceRequest;
+  }
+
+  public Map<ServerRoutingInstance, InstanceRequest> getRequestMap(long requestId, String brokerId, boolean preferTls) {
```

Comment:
The can be simplified with empty init of the request map and adding the request map using putAll.

Reply:
I didnt do that to avoid creating objects. The current code also uses null checks intsead of empty. Also this code is in performance critical path, I've tried to avoid creating objects.